extensions/api/data-management/contractdefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractdefinition/service/ContractDefinitionServiceImpl.java

```diff
@@ -0,0 +1,101 @@
+import org.eclipse.dataspaceconnector.spi.contract.offer.store.ContractDefinitionStore;
+import org.eclipse.dataspaceconnector.spi.query.Criterion;
+import org.eclipse.dataspaceconnector.spi.query.QuerySpec;
+import org.eclipse.dataspaceconnector.spi.transaction.TransactionContext;
+import org.eclipse.dataspaceconnector.spi.types.domain.contract.offer.ContractDefinition;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+public class ContractDefinitionServiceImpl implements ContractDefinitionService {
+    private final ContractDefinitionStore store;
+    private final ContractDefinitionLoader loader;
+    private final TransactionContext transactionContext;
+    private final ContractNegotiationStore contractNegotiationStore;
+
+    public ContractDefinitionServiceImpl(ContractDefinitionStore store, ContractDefinitionLoader loader, TransactionContext transactionContext, ContractNegotiationStore contractNegotiationStore) {
+        this.store = store;
+        this.loader = loader;
+        this.transactionContext = transactionContext;
+        this.contractNegotiationStore = contractNegotiationStore;
+    }
+
+    @Override
+    public ContractDefinition findbyId(String contractDefinitionId) {
+        var querySpec = QuerySpec.Builder.newInstance()
+                .filter(List.of(new Criterion("id", "=", contractDefinitionId)))
+                .build();
+
+        return store.findAll(querySpec).findFirst().orElse(null);
+    }
+
+    @Override
+    public Collection<ContractDefinition> query(QuerySpec query) {
+        return store.findAll(query).collect(toList());
+    }
+
+    @Override
+    public ServiceResult<ContractDefinition> create(ContractDefinition contractDefinition) {
+        var result = new AtomicReference<ServiceResult<ContractDefinition>>();
+
+        transactionContext.execute(() -> {
+            if (findbyId(contractDefinition.getId()) == null) {
+                loader.accept(contractDefinition);
+                result.set(ServiceResult.success(contractDefinition));
+            } else {
+                result.set(ServiceResult.conflict(format("ContractDefinition %s cannot be created because it already exist", contractDefinition.getId())));
+            }
+        });
+
+        return result.get();
+    }
+
+    @Override
+    public ServiceResult<ContractDefinition> delete(String contractDefinitionId) {
+        var result = new AtomicReference<ServiceResult<ContractDefinition>>();
+
+        transactionContext.execute(() -> {
+            var filter = format("contractAgreement.id like %s:%%", contractDefinitionId);
+            var query = QuerySpec.Builder.newInstance().filter(filter).build();
+
+            var negotiationsOnAsset = contractNegotiationStore.queryNegotiations(query);
```

Comment:
why not use `contractNegotiationStore.findContractAgreement(contractDefinitionId)`?

Reply:
Because the contract agreement id is composed by the contract definition id plus a random generated UUID, so using that method will not return the agreement